### ISSUE TITLE
lua coroutine: runtime error: handler.lua:152: attempt to index local 'userLoginInfo' (a userdata value)

### ISSUE THREAD

**Issue by DharaSingh55** [Kong/kong#3591]
I have simple plugin which intercepts all the requests, make a call to redis and gets user data based on access token.
local authTokenJson = cjson.decode(authToken)
auth token is header passed in request, as a json
local userLoginInfo = getUserLoginInfoByRedisCache(conf, accesstoken)

after that i am modifying userLoginInfo values from api's headers and getting lua coroutine: runtime error for almost 5% requests.

 152 userLoginInfo.apiVersion = authTokenJson.apiVersion
153 userLoginInfo.appVersion = authTokenJson.appVersion
154 userLoginInfo.appOs = authTokenJson.osVersion
155 userLoginInfo.deviceId = authTokenJson.deviceId
156 userLoginInfo.userSecureKey  = userSecureKey
157 userLoginInfo.userLocale = authTokenJson.userLocale
158 userLoginInfo.activeDeviceId = authTokenJson.deviceId

**Comment by bungle** [comment]
Do you check the return value?

**Comment by DharaSingh55** [comment]
local authTokenJson = cjson.decode(authToken)
    local deviceId = authTokenJson['deviceId']
    local accesstoken = authTokenJson['accesstoken']

   
    -- fetch data on cache
    local userLoginInfo = getUserLoginInfoByRedisCache(conf, accesstoken)
    if (not userLoginInfo) or err then
        local err
        userLoginInfo, err = getUserLoginInfoByAuthService(conf, deviceId, accesstoken)
        if err then
            return
        end
        if isempty(userLoginInfo) or (userLoginInfo == ngx.null) then
            return responses.send(200, {["status"]="FAIL",["statusCode"]=1006,["message"]="Invalid credentials",["desc"]="Invalid credentials",["debug"]=nil,["exception"]=nil,["data"]=nil})
        end
    end

**Comment by DharaSingh55** [comment]
anything missing here ?

**Comment by DharaSingh55** [comment]
have changed like this 
for k,v in pairs(authTokenJson) do userLoginInfo[k] = v end

still getting the same error
any help here ?

**Comment by bungle** [comment]
Maybe ` cjson.decode(authToken)` fails? Do you check that?

**Comment by hishamhm** [comment]
Since the error message says `attempt to index local 'userLoginInfo' (a userdata value)` and that was a json-decoded value, my guess is that the resulting userdata is `cjson.null`. Note that when you decode JSON, `null` values become `cjson.null` and not Lua `nil`:

```
$ resty -e 'cjson = require"cjson"; v = cjson.decode([[{ "x": null }]]); print(v.x, " ", type(v.x), " ", v.x == cjson.null)'
null userdata true
```

It's hard to follow the logic from your snippets above, but if you are getting `attempt to index ... (a userdata value)`, then most certainly there is a check for `cjson.null` (which is the same as `ngx.null`) missing somewhere.

**Comment by hishamhm** [comment]
I'm closing the issue here as this bug tracker is aimed at tracking Kong bugs. @DharaSingh55 if you believe it's a bug in Kong and not the custom plugin code, feel free to reopen this. Otherwise, we're happy to continue the conversation at https://discuss.konghq.com !
